wordpress-mobile/WordPress-iOS
# [VoiceOver] NoResultsView enhancements

Refs #11487 and #12906. 

This changes the `NoResultsViewController` so that the full view is accessible instead of just the title and subtitle as suggested in #11487. This makes the “no results” message easily discoverable by VoiceOver users. They do not need to have their finger travel far to _find_ the message.

The title and subtitle are now also dictated together. 

Before | After 
--------|-------
![IMG_1041](https://user-images.githubusercontent.com/198826/68789827-6b1db900-0603-11ea-84a3-407a1a3c27ae.PNG)       |       ![IMG_1044](https://user-images.githubusercontent.com/198826/68789821-6b1db900-0603-11ea-857a-7c79b52cb0f3.PNG)
![IMG_1042](https://user-images.githubusercontent.com/198826/68789824-6b1db900-0603-11ea-9b7e-bbb9328d01db.PNG)  | ![IMG_1043](https://user-images.githubusercontent.com/198826/68789822-6b1db900-0603-11ea-993b-1ab8c5c2916a.PNG)

Some attempts at announcing the “no results” message were done to no great results. More information about that starting at https://github.com/wordpress-mobile/WordPress-iOS/issues/11487#issuecomment-551319114 if you're interested. 🙂 

## View Changes

To make this work, I had to modify `NoResults.storyboard`:

- Made `noResultsView` cover the whole Safe Area
- Moved the `noResultsView` alignment and margins to the main stack view

![image](https://user-images.githubusercontent.com/198826/68789689-2e51c200-0603-11ea-99ac-f48aed18e7f7.png)

I tried to create commits for every single constraint change. I hope that helps. 

## Reader → Saved Posts

The `ReaderSavedPostsViewController` usage of `NoResultsVC` was causing issues. Even before this change, it seems that everything under `NoResultsVC.view` was inaccessible. The effect is that VoiceOver would dictate “empty list” instead of the desired “No results” dictation. 

It took me a while but it turned out that the fix was to move the routine for making `NoResultsVC` a child of `ReaderSavedPostsVC` from `viewDidAppear` to `viewDidLoad` (8c452000d864acaf1a18cae17a8e788affcbc651). With this change, the `tableView` is no longer an accessible element and all the intended views underneath `NoResultsVC` are accessible. 

## Outstanding Issues

As described in #11487, VoiceOver ignores the _bookmark icon_ when dictating the subtitle in Reader → Saved Posts' no-results view. I'm not quite sure how to fix this yet. I'll fix it in a separate PR. 

## Testing

Testing involves using VoiceOver on the places where `NoResultsViewController` is used. The following steps is a suggestion: 

1. Turn VoiceOver on. 
2. Navigate to an empty tab in Notifications and validate that the changes described above work correctly. 
3. Navigate to Site → Post List. Enter a string in the search bar that will result in an empty result. Validate the changes described above. 
4. Turn the Internet connection off.
5. Navigate back to an empty tab in Notifications. The tab should show an “Unable to load this page right now.” message. Validate the changes described above. 

## Submitter Checklist
